src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -729,153 +728,277 @@ func (s *service) FetchTagged(tctx thrift.Context, req *rpc.FetchTaggedRequest)
+	return newFetchTaggedResultsIter(&fetchTaggedResultsIterOpts{
+		queryResult: &queryResult,
+		queryOpts:   &opts,
+		fetchData:   fetchData,
+		db:          db,
+		docReader:   docs.NewEncodedDocumentReader(),
+		nsID:        ns,
+		tagEncoder:  tagEncoder,
+		iOpts:       s.opts.InstrumentOptions(),
+	}), nil
+}
+
+// FetchTaggedResultsIter iterates over the results from FetchTagged
+// The iterator is not thread safe and must only be accessed from a single goroutine.
+type FetchTaggedResultsIter interface {
+
+	// NumIDs returns the total number of series IDs in the result.
+	NumIDs() int
+
+	// Exhaustive returns true if NumIDs is all IDs that the query could have returned.
+	Exhaustive() bool
+
+	// HasNextIDIter returns true if there is another series ID to process.
+	HasNextIDIter() bool
+
+	// NextIDIter returns an iterator to process the results for a series ID.
+	// HasNextIDIter must be called before each call of this method.
+	NextIDIter(ctx context.Context) (IDIter, error)
+}
+
+// IDIter iterates over the results for a single series ID from FetchTagged.
+// The iterator is not thread safe and must only be accessed from a single goroutine.
+type IDIter interface {
+
+	// IDResult returns a partial FetchTaggedIDResult without the segments. The segments must be fetched separately with
+	// NextSegments.
+	IDResult() (*rpc.FetchTaggedIDResult_, error)
+
+	// HasNextSegments returns true if there are more Segments to process.
+	HasNextSegments() bool
+
+	// NextSegments returns the next Segments.
+	// HasNextSegments must be called before each call of this method.
+	NextSegments(ctx context.Context) (*rpc.Segments, error)
+}
+
+type idIter struct {
+	docReader  *docs.EncodedDocumentReader
+	nsID       ident.ID
+	tagEncoder serialize.TagEncoder
+	iOpts      instrument.Options
+
+	segments []*rpc.Segments
+	idResult *idResult
+}
+
+func (i *idIter) IDResult() (*rpc.FetchTaggedIDResult_, error) {
+	if i.idResult == nil {
+		return nil, errors.New("reset never called")
 	}
-	if err := s.fetchReadEncoded(ctx, db, response, results, ns, nsIDBytes, callStart, opts, fetchData, encodedDataResults); err != nil {
-		s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
+	metadata, err := docs.MetadataFromDocument(i.idResult.queryResult.Value(), i.docReader)
+	if err != nil {
 		return nil, err
 	}
-
-	// Step 2: If fetching data read the results of the asynchronous block readers.
-	if fetchData {
-		if err := s.fetchReadResults(ctx, response, ns, encodedDataResults); err != nil {
-			s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
-			return nil, err
-		}
+	tags := idxconvert.ToSeriesTags(metadata, idxconvert.Opts{NoClone: true})
+	encodedTags, err := encodeTags(i.tagEncoder, tags, i.iOpts)
+	if err != nil { // This is an invariant, should never happen
+		return nil, tterrors.NewInternalError(err)
 	}
+	tagBytes := make([]byte, len(encodedTags.Bytes()))
+	copy(tagBytes, encodedTags.Bytes())
```

Comment:
@robskillington had to copy the bytes cause the same tag encoder is used for the entire request.